doc/developer/design/20220718_envelopes.md

```diff
@@ -0,0 +1,208 @@
+	to the committed result of some transaction in the upstream database.
+
+**Hard Requirement**: Materialize is _robust_: although the result
+of ingesting arbitrarily corrupted data is unspecified, the damage
+should be scoped to the affected set of sources. It should not cause Materialize to crash,
+nor should it cause general undefined behavior (e.g., incorrect results when querying
+unrelated sources or views).
+
+## Non-Goals
+
+It is a permanent non-goal
+to support arbitrarily corrupted data, or to support the classic `DEBEZIUM` envelope in
+situations that do not fit the constraints described above. The famous quote by
+Charles Babbage applies here.
+
+It is a permanent non-goal to support the exotic custom deduplication options
+that formerly existed for `DEBEZIUM` sources, like `full` or `full_in_range`.
+These were designed to account for the specific requirements of one particular
+customer, and were never documented nor widely supported or used; furthermore, nobody
+presently working at Materialize understands their intended semantics in detail.
+Although the one customer they were designed for is indeed important, we
+were happily able to confirm recently that they are able to use `DEBEZIUM UPSERT`
+for all of their topics, making the issue moot.
+
+It is _currently_ a non-goal to support Microsoft SQL Server in the classic `DEBEZIUM` envelope,
+due to [this issue](https://issues.redhat.com/browse/DBZ-3375) preventing us from
+precisely duplicating records. If we come up with a way to deduplicate records that we're
+confident in, we will re-evaluate this. Note that Debezium data from
+Microsoft SQL Server should work
+fine with `DEBEZIUM UPSERT`
+
+Solving the issue of [bounded input reliance](https://github.com/MaterializeInc/materialize/issues/13534)
+is outside the scope of this project; until that issue is solved, we must assume
+that data is not being compacted away while we're busy reading it, which is
+somewhat problematic as we have no way to express our frontiers back upstream to
+the user's Kafka installation.
+
+Out-of-core scaling of the O(n)-memory envelopes (`UPSERT` and `UPSERT DEBEZIUM`)
+is a future goal, but outside the scope of this document. This will probably involve relying
+on the state kept in `persist`, probably with a local caching layer for frequently-used
+keys on top.
+
+## Description
+
+For `UPSERT` and `DEBEZIUM UPSERT`, we continue maintaining, as we do now,
+an in-memory hash map of the entire state of the relation (keys and values). We use
+this hash map to discover when a new record is an update (as opposed to an insert)
+and to emit a retraction of the old value.
+
+The main new requirement is that on restart, we must hydrate this state by
+reading the old committed value of the collection from `persist`. This is simple
+to do with `persist`'s current APIs.
+
+For the classic `DEBEZIUM` envelope, we must validate that the topic meets the constraints
```

Comment:
this is perhaps much simpler when not dealing with the transaction metadata topic: trying to figure out how to correctly (read: definitely) timestamp errors was something we punted on when starting to use the metadata topic because we didn't always have a tx metadata message to correspond each data message to (e.g. we know there's a message but it's corrupt and we can't read the tx id it's tagged with).  Without a metadata topic, you just timestamp everything as usual.

Doesn't seem like it's imminently pressing but did want to flag